packages/dds/matrix/src/permutationvector.ts

```diff
@@ -150,6 +152,17 @@ export class PermutationVector extends Client {
             new PermutationSegment(length));
     }
 
+    public insertRelative(segment: ISegment, length: number) {
+        const inserted = new PermutationSegment(length);
+
+        return {
+            op: this.insertAtReferencePositionLocal(
+                    new LocalReference(this, segment, /* offset: */ 0, ReferenceType.Transient),
+                    inserted),
+            inserted,
+        };
```

Comment:
During undo/redo, row/col segments are re-inserted at the position of the original segments in the MergeTree.

When a range of rows/cols that have been split into multiple segments, the MergeTree will resolve all of the original segments to the same position (because they have length zero).

Reinserting the segments in reverse order is maybe good enough, but I noticed Tony uses insertAtRefPosition to insert at the exact position of the original segment.

Reply:
What happens in the case of something like:

z's are deleted: xxxxzzzzxxxx
remote insert prior to the delete splits z's: xxxxzzyzzxxxx
undo deletion of z's which will be an insert at pos 4, length 4.  Do they all come before the the y?

Maybe this is not a valid interaction or I'm missing something?

Reply:
I think you're right.  I opened #3964 to track.

Reply:
There could also be other deletes inter-mixed, so exact positioning is important to not mess up other operations and their order. i made a change to add per-position local references, that also include optional before/after tombstones. Eventually i plan to move undo redo of delete to use these, rather than depending on segment order which creates issues for breakTie

---
In reply to: [505865768](https://github.com/microsoft/FluidFramework/pull/3946#discussion_r505865768) [](ancestors = 505865768)

Reply:
this is expected. basically undo/redo works from the current local perspective. This can lead to some localized behaviors that isn't always expected, but is still valid in regards to the merge tree. so far we haven't see users have a problem with this, as tight interleaving of operations isn't something we need a lot of in shared string

---
In reply to: [506431314](https://github.com/microsoft/FluidFramework/pull/3946#discussion_r506431314) [](ancestors = 506431314)